File: Sorting/MergeSort.cpp
#include<iostream>
using namespace std;
void merge(int arr[],int l,int m,int r){
    int p = l,q = m+1,k=0;
    int A[r-l+1];
    for(int i=0;i<=r;i++){
        if(p>m)
            A[k++] = arr[q++];
        else if(q>r)
            A[k++] = arr[p++];
        else if(arr[p]<arr[q])
            A[k++] = arr[p++];
        else
            A[k++] = arr[q++];
    }
    for(int i=0;i<k;i++)
        arr[l++] = A[i];
}
void mergeSort(int arr[],int l, int r){
    int mid;
    if(l<r){
        mid = (l+r)/2;
        mergeSort(arr,l,mid);
        mergeSort(arr,mid+1,r);
        merge(arr,l,mid,r);
    }
}
int main(){
    int n;
    cin>>n;
    int arr[n];
    for(int i=0;i<n;i++){
        cin>>arr[i];
    }
    cout<<endl<<"Entered array is : ";
    for(int i=0;i<n;i++){
        cout<<arr[i]<<" ";
    }
    mergeSort(arr,0,n-1);
    cout<<endl<<"Sorted array is : ";
    for(int i=0;i<n;i++){
        cout<<arr[i]<<" ";
    }
}

File: Sorting/MergeSort.py
def merge(arr,l,mid,r):
    p = l
    q = mid + 1
    k = 0
    A = [0] * (r-l+1) #(intialise point * length)This is how u initialise list
    # print(A)
    for i in range(l,r+1):
        if p>mid:
            A[k] = arr[q]
            k += 1
            q += 1
            
        elif(q>r):
            A[k] = arr[p]
            k += 1
            p += 1
            
        elif(arr[p]<arr[q]):
            A[k] = arr[p]
            k += 1
            p += 1
            
        else:
            A[k] = arr[q]
            k += 1
            q += 1
    
    for i in range(0,k):
        arr[l] = A[i]
        l += 1

def mergeSort(arr , l , r):
    if l<r:
        mid = (l+r)//2
        mergeSort(arr,l,mid)
        mergeSort(arr,mid+1,r)
        merge(arr,l,mid,r)
    return

def main():
    num = int(input())
    
    '''
        Method1:
        1
        2
        3
        4
        5
    # arr = []
    # arr = list()
    
    # for i in range(0,num):
    #     tmp = int(input())
    #     arr.append(tmp)
    
    '''
    
    # Method2:  1 2 3 4 5
    arr = input().split(" ")
    for i in range(0,num):
        arr[i] = int(arr[i])
    mergeSort(arr,0,num-1)
    print('After sorting the array is : ')
    for i in range(0,num):
        print(arr[i],end='\n')

main()